// @looker/components/lib/Divider/Divider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Divider = void 0;

var _designTokens = require("@looker/design-tokens");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var appearanceVariant = (0, _styledSystem.variant)({
  prop: 'appearance',
  variants: {
    dark: {
      bg: 'ui4'
    },
    default: {
      bg: 'ui3'
    },
    light: {
      bg: 'ui2'
    },
    onDark: {
      bg: 'text4'
    }
  }
});

var Divider = _styledComponents.default.hr.attrs(props => ({
  bg: props.customColor
})).withConfig({
  displayName: "Divider",
  componentId: "sc-1ceogl5-0"
})(["", " ", " ", " border:none;height:", ";width:100%;", ""], _designTokens.reset, _designTokens.position, _designTokens.space, props => props.size, props => props.customColor ? _designTokens.color : appearanceVariant);

exports.Divider = Divider;
Divider.defaultProps = {
  appearance: 'default',
  size: '1px'
};
