// @looker/components/lib/Button/ButtonItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonItem = exports.buttonItemHeight = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _designTokens = require("@looker/design-tokens");

var _ButtonSetContext = require("./ButtonSetContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ButtonLayout = (0, _react.forwardRef)((_ref, ref) => {
  var {
    children,
    onClick,
    value
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children", "onClick", "value"]);

  var {
    disabled,
    value: contextValue,
    onItemClick
  } = (0, _react.useContext)(_ButtonSetContext.ButtonSetContext);

  function handleClick(e) {
    onClick && onClick(e);

    if (!e.defaultPrevented) {
      onItemClick && onItemClick(e);
    }
  }

  var itemValue = value !== undefined ? value : typeof children === 'string' ? children : '';
  var selected = contextValue ? typeof contextValue === 'string' ? contextValue === itemValue : contextValue.includes(itemValue) : false;
  return _react.default.createElement("button", _extends({
    type: "button",
    "aria-pressed": selected,
    ref: ref,
    onClick: handleClick,
    value: itemValue,
    disabled: disabled
  }, (0, _designTokens.omitStyledProps)(props)), children);
});
ButtonLayout.displayName = 'ButtonLayout';
var buttonItemHeight = 36;
exports.buttonItemHeight = buttonItemHeight;
var ButtonItem = (0, _styledComponents.default)(ButtonLayout).withConfig({
  displayName: "ButtonItem",
  componentId: "sc-22szay-0"
})(["align-items:center;background:transparent;border:none;color:", ";cursor:pointer;display:inline-flex;font-size:", ";height:", "px;justify-content:center;margin:0;padding:0 ", ";transition:background ", " ease;user-select:none;&[aria-pressed='false']:not([disabled]):hover{background:", ";}&:focus{box-shadow:0 0 0.5px 1px ", ";outline:none;}&[disabled]{color:", ";cursor:default;}&[aria-pressed='true']{background:", ";border-color:", ";color:", ";text-shadow:-0.025ex 0 currentColor,0.025ex 0 currentColor;&[disabled]{color:", ";}}", " ", ""], (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.colors.text3;
}, (_ref3) => {
  var {
    theme
  } = _ref3;
  return theme.fontSizes.small;
}, buttonItemHeight, (_ref4) => {
  var {
    theme
  } = _ref4;
  return theme.space.small;
}, (_ref5) => {
  var {
    theme
  } = _ref5;
  return theme.transitions.durationQuick;
}, (_ref6) => {
  var {
    theme
  } = _ref6;
  return theme.colors.keySubtle;
}, (_ref7) => {
  var {
    theme
  } = _ref7;
  return theme.colors.keyFocus;
}, props => props.theme.colors.text5, (_ref8) => {
  var {
    theme
  } = _ref8;
  return theme.colors.keyAccent;
}, (_ref9) => {
  var {
    theme
  } = _ref9;
  return theme.colors.keyAccent;
}, (_ref10) => {
  var {
    theme
  } = _ref10;
  return theme.colors.key;
}, (_ref11) => {
  var {
    theme
  } = _ref11;
  return theme.colors.keyFocus;
}, _designTokens.space, _designTokens.typography);
exports.ButtonItem = ButtonItem;
