// @looker/components/lib/Button/size.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buttonSize = exports.buttonSizeMap = void 0;

var _styledSystem = require("styled-system");

var buttonSizeMap = {
  xxsmall: 20,
  xsmall: 24,
  small: 28,
  medium: 36,
  large: 44
};
exports.buttonSizeMap = buttonSizeMap;
var buttonSize = (0, _styledSystem.variant)({
  prop: 'size',
  variants: {
    xxsmall: {
      fontSize: 'xxsmall',
      height: "".concat(buttonSizeMap.xxsmall, "px"),
      px: 'xsmall'
    },
    xsmall: {
      fontSize: 'xxsmall',
      height: "".concat(buttonSizeMap.xsmall, "px"),
      px: 'small'
    },
    small: {
      fontSize: 'xsmall',
      height: "".concat(buttonSizeMap.small, "px"),
      px: 'large'
    },
    medium: {
      fontSize: 'small',
      height: "".concat(buttonSizeMap.medium, "px"),
      px: '24px'
    },
    large: {
      fontSize: 'large',
      height: "".concat(buttonSizeMap.large, "px"),
      px: 'xlarge'
    }
  }
});
exports.buttonSize = buttonSize;
