// @looker/components/lib/Form/Fields/Field.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Field = exports.omitFieldProps = exports.pickFieldProps = exports.fieldPropKeys = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _omit = _interopRequireDefault(require("lodash/omit"));

var _pick = _interopRequireDefault(require("lodash/pick"));

var _Text = require("../../Text");

var _Fieldset = require("../Fieldset");

var _InputText = require("../Inputs/InputText/InputText");

var _Label = require("../Label");

var _VisuallyHidden = require("../../VisuallyHidden");

var _ValidationMessage = require("../ValidationMessage");

var _RequiredStar = require("./RequiredStar");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var fieldPropKeys = ['description', 'detail', 'id', 'inline', 'label', 'hideLabel', 'labelWidth', 'validationMessage', 'width'];
exports.fieldPropKeys = fieldPropKeys;

var pickFieldProps = props => (0, _pick.default)(props, [...fieldPropKeys, 'disabled', 'required', 'className', 'autoResize']);

exports.pickFieldProps = pickFieldProps;

var omitFieldProps = props => (0, _omit.default)(props, fieldPropKeys);

exports.omitFieldProps = omitFieldProps;

var FieldLayout = (_ref) => {
  var {
    className,
    children,
    description,
    detail,
    id,
    label,
    hideLabel,
    required,
    validationMessage
  } = _ref;
  var {
    fieldsHideLabel
  } = (0, _react.useContext)(_Fieldset.FieldsetContext);

  var fieldDescription = description && _react.default.createElement(_Text.Paragraph, {
    mt: "xsmall",
    fontSize: "xsmall",
    color: "text4"
  }, description);

  var fieldValidation = validationMessage && _react.default.createElement(_ValidationMessage.ValidationMessage, validationMessage);

  var labelComponent = _react.default.createElement(_Label.Label, {
    htmlFor: id,
    id: "".concat(id, "-labelledby")
  }, label, required && _react.default.createElement(_RequiredStar.RequiredStar, null));

  return _react.default.createElement("div", {
    className: className
  }, (fieldsHideLabel || hideLabel) && hideLabel !== false ? _react.default.createElement(_VisuallyHidden.VisuallyHidden, null, labelComponent) : labelComponent, detail && _react.default.createElement(FieldDetail, null, detail), _react.default.createElement(InputArea, null, children), _react.default.createElement(MessageArea, {
    id: "".concat(id, "-describedby")
  }, fieldDescription, fieldValidation));
};

var FieldDetail = (0, _styledComponents.default)(_Text.Text).withConfig({
  displayName: "Field__FieldDetail",
  componentId: "m8c1e2-0"
})([""]);
FieldDetail.defaultProps = {
  fontSize: 'xsmall'
};

var InputArea = _styledComponents.default.div.withConfig({
  displayName: "Field__InputArea",
  componentId: "m8c1e2-1"
})([""]);

var MessageArea = _styledComponents.default.div.withConfig({
  displayName: "Field__MessageArea",
  componentId: "m8c1e2-2"
})([""]);

var fieldLabelCSS = inline => inline ? (0, _styledComponents.css)(["height:", ";justify-self:end;line-height:", ";padding-right:", ";text-align:right;"], _InputText.inputHeight, _InputText.inputHeight, (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.space.small;
}) : (0, _styledComponents.css)(["line-height:", ";padding-bottom:", ";"], (_ref3) => {
  var {
    theme
  } = _ref3;
  return theme.lineHeights.xsmall;
}, (_ref4) => {
  var {
    theme
  } = _ref4;
  return theme.space.xsmall;
});

var Field = (0, _styledComponents.default)(FieldLayout).withConfig({
  displayName: "Field",
  componentId: "m8c1e2-3"
})(["align-items:left;display:", ";grid-template-areas:", ";grid-template-columns:", ";height:fit-content;justify-content:space-between;width:", ";", "{align-items:center;", " grid-area:input;}", "{grid-area:messages;}& > ", "{grid-area:label;", "}", "{grid-area:detail;justify-self:end;padding-left:", ";", "}", "{margin-top:", ";}"], (_ref5) => {
  var {
    autoResize
  } = _ref5;
  return autoResize ? 'inline-grid' : 'grid';
}, (_ref6) => {
  var {
    inline
  } = _ref6;
  return inline ? '"label input detail" ". messages messages"' : '"label detail" "input input" "messages messages"';
}, (_ref7) => {
  var {
    inline
  } = _ref7;
  return inline ? '150px 1fr' : undefined;
}, (_ref8) => {
  var {
    autoResize,
    width
  } = _ref8;
  return width || autoResize ? 'fit-content' : '100%';
}, InputArea, (_ref9) => {
  var {
    autoResize
  } = _ref9;
  return autoResize && (0, _styledComponents.css)(["align-items:stretch;display:flex;flex-direction:column;"]);
}, MessageArea, _Label.Label, (_ref10) => {
  var {
    inline
  } = _ref10;
  return fieldLabelCSS(inline);
}, FieldDetail, (_ref11) => {
  var {
    theme: {
      space
    }
  } = _ref11;
  return space.small;
}, (_ref12) => {
  var {
    inline
  } = _ref12;
  return inline && (0, _styledComponents.css)(["align-self:center;"]);
}, _ValidationMessage.ValidationMessage, (_ref13) => {
  var {
    theme
  } = _ref13;
  return theme.space.xsmall;
});
exports.Field = Field;
