// @looker/components/lib/Form/Fields/FieldInline.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FieldInline = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Label = require("../Label/Label");

var _ValidationMessage = require("../ValidationMessage/ValidationMessage");

var _RequiredStar = require("./RequiredStar");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FieldInlineLayout = (_ref) => {
  var {
    className,
    children,
    detail,
    label,
    id,
    required,
    validationMessage
  } = _ref;
  return _react.default.createElement("label", {
    className: className,
    htmlFor: id
  }, _react.default.createElement(_Label.Label, {
    as: "span"
  }, label, required && _react.default.createElement(_RequiredStar.RequiredStar, null)), detail && _react.default.createElement(FieldDetail, null, detail), _react.default.createElement(InputArea, null, children), _react.default.createElement(MessageArea, {
    id: "".concat(id, "-describedby")
  }, validationMessage ? _react.default.createElement(_ValidationMessage.ValidationMessage, validationMessage) : null));
};

var InputArea = _styledComponents.default.div.withConfig({
  displayName: "FieldInline__InputArea",
  componentId: "kl4qjf-0"
})([""]);

var FieldDetail = _styledComponents.default.div.withConfig({
  displayName: "FieldInline__FieldDetail",
  componentId: "kl4qjf-1"
})([""]);

var MessageArea = _styledComponents.default.div.withConfig({
  displayName: "FieldInline__MessageArea",
  componentId: "kl4qjf-2"
})([""]);

var FieldInline = (0, _styledComponents.default)(FieldInlineLayout).withConfig({
  displayName: "FieldInline",
  componentId: "kl4qjf-3"
})(["align-items:center;display:grid;grid-template-areas:'input label detail' '. messages messages';grid-template-columns:repeat(2,max-content) 1fr;line-height:", ";", "{grid-area:input;}", "{align-items:center;color:", ";display:flex;font-size:", ";font-weight:normal;grid-area:label;padding-left:", ";}", "{align-content:center;display:flex;grid-area:detail;justify-content:flex-end;margin-left:", ";}", "{grid-area:messages;padding-left:", ";}"], (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.lineHeights.small;
}, InputArea, _Label.Label, (_ref3) => {
  var {
    theme,
    disabled
  } = _ref3;
  return disabled && theme.colors.text5;
}, (_ref4) => {
  var {
    theme
  } = _ref4;
  return theme.fontSizes.small;
}, (_ref5) => {
  var {
    theme
  } = _ref5;
  return theme.space.xsmall;
}, FieldDetail, (_ref6) => {
  var {
    theme: {
      space
    }
  } = _ref6;
  return space.xxsmall;
}, MessageArea, (_ref7) => {
  var {
    theme
  } = _ref7;
  return theme.space.small;
});
exports.FieldInline = FieldInline;
