// @looker/components/lib/Calendar/calendar-size.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.calendarSpacing = exports.calendarSize = exports.daySizeMap = void 0;

var _styledSystem = require("styled-system");

var daySizeMap = {
  small: 32,
  medium: 36,
  large: 44
};
exports.daySizeMap = daySizeMap;
var calendarSize = (0, _styledSystem.variant)({
  prop: 'size',
  variants: {
    small: {
      fontSize: 'xsmall',
      height: "".concat(daySizeMap.small, "px"),
      width: "".concat(daySizeMap.small, "px")
    },
    medium: {
      fontSize: 'small',
      height: "".concat(daySizeMap.medium, "px"),
      width: "".concat(daySizeMap.medium, "px")
    },
    large: {
      fontSize: 'medium',
      height: "".concat(daySizeMap.large, "px"),
      width: "".concat(daySizeMap.large, "px")
    }
  }
});
exports.calendarSize = calendarSize;
var calendarSpacing = (0, _styledSystem.variant)({
  prop: 'size',
  variants: {
    small: {
      p: 'small'
    },
    medium: {
      p: 'medium'
    },
    large: {
      p: 'large'
    }
  }
});
exports.calendarSpacing = calendarSpacing;
