// @looker/components/lib/utils/useMeasuredElement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMeasuredElement = void 0;

var _react = require("react");

var _resizeObserverPolyfill = _interopRequireDefault(require("resize-observer-polyfill"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function measureElement(element) {
  if (!element) {
    return {
      bottom: 0,
      height: 0,
      left: 0,
      right: 0,
      top: 0,
      width: 0
    };
  }

  return element.getBoundingClientRect();
}

var useMeasuredElement = element => {
  var [rect, setRect] = (0, _react.useState)(measureElement());
  var handleResize = (0, _react.useCallback)(() => {
    element && setRect(measureElement(element));
  }, [element]);
  (0, _react.useLayoutEffect)(() => {
    if (!element) {
      return;
    }

    handleResize();
    var resizeObserver = new _resizeObserverPolyfill.default(() => handleResize());

    if (element) {
      resizeObserver.observe(element);
    }

    window.addEventListener('resize', handleResize);
    return () => {
      if (!resizeObserver) {
        return;
      }

      resizeObserver.disconnect();
      window.removeEventListener('resize', handleResize);
    };
  }, [handleResize, element]);
  return rect;
};

exports.useMeasuredElement = useMeasuredElement;
