// @looker/components/lib/Form/Inputs/TextArea/TextArea.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextArea = void 0;

var _pick = _interopRequireDefault(require("lodash/pick"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _InputText = require("../InputText");

var _Icon = require("../../../Icon");

var _simple = require("../../../Layout/utils/simple");

var _InputProps = require("../InputProps");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var TextAreaLayout = (_ref) => {
  var {
    className,
    validationType
  } = _ref,
      props = _objectWithoutProperties(_ref, ["className", "validationType"]);

  var textareaProps = (0, _pick.default)(props, _InputProps.inputPropKeys);
  return _react.default.createElement("div", {
    className: className
  }, _react.default.createElement("textarea", _extends({
    "aria-invalid": validationType === 'error' ? 'true' : undefined
  }, textareaProps)), validationType && _react.default.createElement(_Icon.Icon, {
    name: "CircleInfo",
    color: "critical",
    size: 20
  }));
};

var TextAreaResize = resize => resize === false ? 'none' : resize === true ? 'vertical' : resize;

var TextArea = (0, _styledComponents.default)(TextAreaLayout).withConfig({
  displayName: "TextArea",
  componentId: "sc-10ezzv1-0"
})(["height:fit-content;position:relative;width:100%;", "{position:absolute;right:0.5rem;top:0.5rem;}textarea{", " ", " padding:", ";padding-right:", ";resize:", ";width:100%;&:hover{", "}&:focus,:focus-within{", "}", " ", "}"], _Icon.Icon, _simple.simpleLayoutCSS, _InputText.inputCSS, (_ref2) => {
  var {
    theme
  } = _ref2;
  return "".concat(theme.space.xsmall, " ").concat(theme.space.small);
}, props => props.theme.space.xlarge, props => TextAreaResize(props.resize), _InputText.inputTextHover, _InputText.inputTextFocus, props => props.disabled ? _InputText.inputTextDisabled : '', _InputText.inputTextValidation);
exports.TextArea = TextArea;
TextArea.defaultProps = {
  minHeight: '6.25rem',
  resize: 'vertical'
};
TextArea.displayName = 'TextArea';
