// @looker/components/lib/Pagination/Pagination.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pagination = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Layout = require("../Layout");

var _Button = require("../Button");

var _Text = require("../Text");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var PaginationButton = props => _react.default.createElement(_Button.IconButton, _extends({
  outline: true,
  mx: "xxsmall"
}, props));

var PaginationLayout = (_ref) => {
  var {
    className,
    current,
    pages,
    onChange
  } = _ref;
  if (pages <= 1) return null;

  var first = () => onChange(1);

  var previous = () => onChange(current - 1);

  var next = () => onChange(current + 1);

  var last = () => onChange(pages);

  return _react.default.createElement(_Layout.Flex, {
    alignItems: "center",
    className: className
  }, _react.default.createElement(PaginationButton, {
    label: "First page of results",
    icon: "DoubleChevronLeft",
    onClick: first,
    disabled: current <= 2
  }), _react.default.createElement(PaginationButton, {
    label: "Previous page of results",
    icon: "CaretLeft",
    onClick: previous,
    disabled: current === 1
  }), _react.default.createElement(_Text.Text, {
    fontSize: "small",
    mx: "xxsmall"
  }, _react.default.createElement("b", null, current), " of ", pages), _react.default.createElement(PaginationButton, {
    label: "Next page of results",
    icon: "CaretRight",
    onClick: next,
    disabled: pages - current === 0
  }), _react.default.createElement(PaginationButton, {
    mr: "none",
    label: "Last page of results",
    icon: "DoubleChevronRight",
    onClick: last,
    disabled: pages - current <= 1
  }));
};

var Pagination = (0, _styledComponents.default)(PaginationLayout).withConfig({
  displayName: "Pagination",
  componentId: "kh6un3-0"
})([""]);
exports.Pagination = Pagination;
