// @looker/components/lib/List/List.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.List = void 0;

var _designTokens = require("@looker/design-tokens");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var typeVariant = (0, _styledSystem.variant)({
  prop: 'type',
  variants: {
    bullet: {
      listStyleType: 'disc',
      pl: 'medium'
    },
    letter: {
      listStyleType: 'upper-alpha',
      pl: 'medium'
    },
    none: {
      listStyleType: 'none'
    },
    number: {
      listStyleType: 'decimal',
      pl: 'medium'
    }
  }
});

var List = _styledComponents.default.ul.attrs(props => ({
  as: ['letter', 'number'].includes(String(props.type)) ? 'ol' : undefined,
  type: props.nomarker ? 'none' : props.type
})).withConfig({
  displayName: "List",
  componentId: "q1s69i-0"
})(["", " ", " ", " ", " ", " ", ""], _designTokens.reset, _designTokens.typography, typeVariant, _designTokens.space, _designTokens.position, _designTokens.layout);

exports.List = List;
List.defaultProps = {
  type: 'none'
};
