// @looker/components/lib/Portal/Portal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Portal = exports.getPortalRoot = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactDom = require("react-dom");

var _styledComponents = _interopRequireDefault(require("styled-components"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var rootId = 'modal-root';

var getPortalRoot = () => {
  var existing = document.getElementById(rootId);

  if (existing) {
    return existing;
  } else {
    var newElement = document.createElement('div');
    newElement.id = 'modal-root';
    document.body.appendChild(newElement);
    return newElement;
  }
};

exports.getPortalRoot = getPortalRoot;
var Portal = (0, _react.forwardRef)((_ref, ref) => {
  var {
    children
  } = _ref;
  var el = (0, _react.useRef)(document.createElement('div'));
  (0, _react.useEffect)(() => {
    var root = getPortalRoot();
    if (!root) return;
    var elCurrent = el.current;
    root.appendChild(elCurrent);
    return () => {
      root.removeChild(elCurrent);
    };
  }, [el]);

  var content = _react.default.createElement(InvisiBox, {
    ref: ref,
    tabIndex: -1
  }, children);

  return (0, _reactDom.createPortal)(content, el.current);
});
exports.Portal = Portal;
Portal.displayName = 'Portal';

var InvisiBox = _styledComponents.default.div.withConfig({
  displayName: "Portal__InvisiBox",
  componentId: "sc-8jnv99-0"
})(["align-items:center;bottom:0;display:flex;justify-content:center;left:0;pointer-events:none;position:fixed;right:0;top:0;z-index:", ";*{pointer-events:auto;}"], (_ref2) => {
  var {
    theme: {
      zIndexFloor
    }
  } = _ref2;
  return zIndexFloor;
});
