// @looker/components/lib/Spinner/Spinner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Spinner = void 0;

var _designTokens = require("@looker/design-tokens");

var _omit = _interopRequireDefault(require("lodash/omit"));

var _range = _interopRequireDefault(require("lodash/range"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _SpinnerMarker = require("./SpinnerMarker");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SpinnerFactory = props => {
  var {
    color = 'text0',
    markers = 13,
    markerRadius,
    speed = 1000
  } = props;
  return _react.default.createElement(Style, (0, _omit.default)(props, 'color', 'markers', 'markersRadius', 'speed'), (0, _range.default)(markers).map(i => _react.default.createElement(_SpinnerMarker.SpinnerMarker, {
    backgroundColor: color,
    key: i,
    speed: speed,
    markers: markers,
    markerIndex: i,
    markerRadius: markerRadius
  })));
};

var Style = _styledComponents.default.div.withConfig({
  displayName: "Spinner__Style",
  componentId: "dvoyit-0"
})(["", " ", " ", " height:", "px;position:relative;width:", "px;"], _designTokens.reset, _designTokens.space, _designTokens.position, props => props.size, props => props.size);

Style.defaultProps = {
  size: 30
};
var Spinner = (0, _styledComponents.default)(SpinnerFactory).withConfig({
  displayName: "Spinner",
  componentId: "dvoyit-1"
})([""]);
exports.Spinner = Spinner;
