// @looker/components/lib/Menu/MenuGroupLabel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuGroupLabel = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Heading = require("../Text/Heading");

var _MenuGroup = require("./MenuGroup.hooks");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MenuGroupLabel = (_ref) => {
  var {
    children
  } = _ref;
  var labelShimRef = (0, _react.useRef)();
  var isLabelShimVisible = (0, _MenuGroup.useElementVisibility)(labelShimRef);
  return _react.default.createElement(MenuGroupLabelWrapper, {
    renderBoxShadow: !isLabelShimVisible
  }, _react.default.createElement("div", {
    ref: labelShimRef,
    style: {
      height: '0'
    }
  }), _react.default.createElement(_Heading.Heading, {
    fontSize: "small",
    as: "h2",
    pl: "medium",
    fontWeight: "semiBold"
  }, children));
};

exports.MenuGroupLabel = MenuGroupLabel;

var MenuGroupLabelWrapper = _styledComponents.default.div.withConfig({
  displayName: "MenuGroupLabel__MenuGroupLabelWrapper",
  componentId: "a4tegq-0"
})(["background:", ";box-shadow:", ";margin-bottom:", ";position:sticky;top:-1px;", "{color:", ";}"], (_ref2) => {
  var {
    theme: {
      colors
    }
  } = _ref2;
  return colors.background;
}, (_ref3) => {
  var {
    renderBoxShadow,
    theme: {
      colors
    }
  } = _ref3;
  return renderBoxShadow ? "0 4px 8px -2px ".concat(colors.ui2) : 'none';
}, (_ref4) => {
  var {
    theme
  } = _ref4;
  return theme.space.xxsmall;
}, _Heading.Heading, (_ref5) => {
  var {
    theme
  } = _ref5;
  return theme.colors.text2;
});
