// @looker/components/lib/Menu/MenuItemLayout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuItemLayout = exports.MenuItemLayoutGrid = void 0;

var _omit = _interopRequireDefault(require("lodash/omit"));

var _designTokens = require("@looker/design-tokens");

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Icon = require("../Icon");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var MenuItemWrapper = (0, _react.forwardRef)((props, ref) => {
  return _react.default.createElement("li", _extends({}, (0, _omit.default)(props, 'compact', 'focusVisible', 'hasIcon'), {
    ref: ref
  }));
});
MenuItemWrapper.displayName = 'MenuItemWrapper';

var MenuItemLayoutGrid = _styledComponents.default.div.withConfig({
  displayName: "MenuItemLayout__MenuItemLayoutGrid",
  componentId: "na7wpk-0"
})([""]);

exports.MenuItemLayoutGrid = MenuItemLayoutGrid;
var MenuItemLayout = (0, _styledComponents.default)(MenuItemWrapper).withConfig({
  displayName: "MenuItemLayout",
  componentId: "na7wpk-1"
})(["align-items:center;color:", ";display:flex;flex-wrap:wrap;font-size:", ";font-weight:", ";outline:none;text-decoration:none;transition:", ";button,a{", " align-items:center;background:transparent;border:none;color:inherit;cursor:pointer;flex:1;font-size:inherit;font-weight:inherit;height:", ";outline:none;padding:", ";position:relative;text-align:left;text-decoration:none;&:hover,&:focus{color:inherit;text-decoration:none;}", "{align-items:center;display:grid;grid-gap:", ";grid-template-columns:", ";}}", " ", "{color:", ";transition:color ", ";}:hover,&[aria-current='true']{background:", ";color:", ";", "{color:", ";}}&[aria-current='true']{font-weight:", ";}&[disabled]{color:", ";button,a{cursor:not-allowed;}&:hover{background:", ";color:", ";", "{color:", ";}}}"], (_ref) => {
  var {
    theme: {
      colors
    }
  } = _ref;
  return colors.text1;
}, (_ref2) => {
  var {
    theme: {
      fontSizes
    }
  } = _ref2;
  return fontSizes.small;
}, (_ref3) => {
  var {
    theme: {
      fontWeights
    }
  } = _ref3;
  return fontWeights.normal;
}, (_ref4) => {
  var {
    theme: {
      easings,
      transitions
    }
  } = _ref4;
  return "background ".concat(transitions.durationQuick, " ").concat(easings.ease, ",\n    color ").concat(transitions.durationQuick, " ").concat(easings.ease);
}, _designTokens.reset, (_ref5) => {
  var {
    compact
  } = _ref5;
  return compact ? '32px' : '40px';
}, (_ref6) => {
  var {
    compact,
    theme: {
      space: {
        xxsmall,
        xsmall,
        medium
      }
    }
  } = _ref6;
  return compact ? "".concat(xxsmall, " ").concat(medium) : "".concat(xsmall, " ").concat(medium);
}, MenuItemLayoutGrid, (_ref7) => {
  var {
    compact,
    theme: {
      space
    }
  } = _ref7;
  return compact ? space.xsmall : space.small;
}, (_ref8) => {
  var {
    hasIcon
  } = _ref8;
  return hasIcon ? '24px 1fr' : ' 1fr';
}, (_ref9) => {
  var {
    focusVisible,
    theme: {
      colors
    }
  } = _ref9;
  return focusVisible && "&:focus-within button:after,\n  &:focus-within a:after {\n    content: '';\n    display:block;\n    border: solid 2px ".concat(colors.keyFocus, ";\n    border-radius: 2px;\n    margin: 0 1px;\n    position: absolute;\n    top: 0;\n    left: 0;\n    right: 0;\n    bottom: 0;\n  }\n  ");
}, _Icon.Icon, (_ref10) => {
  var {
    theme: {
      colors
    }
  } = _ref10;
  return colors.text6;
}, (_ref11) => {
  var {
    theme: {
      easings,
      transitions
    }
  } = _ref11;
  return "".concat(transitions.durationQuick, " ").concat(easings.ease);
}, (_ref12) => {
  var {
    theme: {
      colors
    }
  } = _ref12;
  return colors.ui1;
}, (_ref13) => {
  var {
    theme: {
      colors
    }
  } = _ref13;
  return colors.text0;
}, _Icon.Icon, (_ref14) => {
  var {
    theme: {
      colors
    }
  } = _ref14;
  return colors.text5;
}, (_ref15) => {
  var {
    theme: {
      fontWeights
    }
  } = _ref15;
  return fontWeights.semiBold;
}, (_ref16) => {
  var {
    theme: {
      colors
    }
  } = _ref16;
  return colors.text5;
}, (_ref17) => {
  var {
    theme: {
      colors
    }
  } = _ref17;
  return colors.background;
}, (_ref18) => {
  var {
    theme: {
      colors
    }
  } = _ref18;
  return colors.text5;
}, _Icon.Icon, (_ref19) => {
  var {
    theme: {
      colors
    }
  } = _ref19;
  return colors.text6;
});
exports.MenuItemLayout = MenuItemLayout;
