// @looker/components/lib/Text/Heading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Heading = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _designTokens = require("@looker/design-tokens");

var _TextBase = require("./TextBase");

var _text_variant = require("./text_variant");

var _truncate = require("./truncate");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var headingLevelFontSize = props => {
  switch (props.as) {
    case 'h1':
      return 'xxlarge';

    case 'h3':
      return 'large';

    case 'h4':
      return 'medium';

    case 'h5':
      return 'small';

    case 'h6':
      return 'xsmall';

    case 'h2':
    default:
      return 'xlarge';
  }
};

var headingLineHeight = props => props.fontSize ? props.fontSize : headingLevelFontSize(props);

var Heading = (0, _styledComponents.default)(_TextBase.TextBase).attrs(props => ({
  fontSize: props.fontSize || headingLevelFontSize(props),
  lineHeight: props.lineHeight || headingLineHeight(props)
})).withConfig({
  displayName: "Heading",
  componentId: "sc-63s0tz-0"
})(["", " ", " ", ""], _designTokens.textTransform, _text_variant.textVariant, _truncate.truncate);
exports.Heading = Heading;
Heading.defaultProps = {
  as: 'h2',
  fontWeight: 'normal',
  variant: 'default'
};
