// @looker/components/lib/Form/Inputs/InputColor/Swatch/Swatch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Swatch = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _designTokens = require("@looker/design-tokens");

var _styledSystem = require("styled-system");

var _InputText = require("../../../Inputs/InputText");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var emptySwatch = "position: relative;\n  &:after {\n    position: absolute;\n    display: block;\n    content: '';\n    height: 1px;\n    width: calc(100% + 12px);\n    background: red;\n    /* left: 50%; */\n    top: 50%;\n    margin-left: -6px;\n    transform: rotate(-45deg);\n  }";

var Swatch = _styledComponents.default.div.withConfig({
  displayName: "Swatch",
  componentId: "sc-1ldlx27-0"
})(["", " ", " ", " ", " background-color:", ";flex-shrink:0;margin-top:auto;&:hover{", "}", " ", ""], _designTokens.reset, _InputText.inputCSS, _styledSystem.width, _styledSystem.height, props => props.color, _InputText.inputTextHover, props => props.color === 'transparent' && emptySwatch, props => props.disabled ? _InputText.inputTextDisabled : '');

exports.Swatch = Swatch;
Swatch.defaultProps = {
  color: 'transparent',
  height: _InputText.inputHeight,
  width: _InputText.inputHeight
};
