// @looker/design-tokens/lib/system/typography/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FontFamilies", {
  enumerable: true,
  get: function get() {
    return _font_families.FontFamilies;
  }
});
Object.defineProperty(exports, "FontFamilyChoices", {
  enumerable: true,
  get: function get() {
    return _font_families.FontFamilyChoices;
  }
});
Object.defineProperty(exports, "FontFamilyProps", {
  enumerable: true,
  get: function get() {
    return _font_families.FontFamilyProps;
  }
});
Object.defineProperty(exports, "FontSizes", {
  enumerable: true,
  get: function get() {
    return _font_sizes.FontSizes;
  }
});
Object.defineProperty(exports, "FontSizeProps", {
  enumerable: true,
  get: function get() {
    return _font_sizes.FontSizeProps;
  }
});
Object.defineProperty(exports, "FontSizeRamp", {
  enumerable: true,
  get: function get() {
    return _font_sizes.FontSizeRamp;
  }
});
Object.defineProperty(exports, "FontWeights", {
  enumerable: true,
  get: function get() {
    return _font_weights.FontWeights;
  }
});
Object.defineProperty(exports, "FontWeightProps", {
  enumerable: true,
  get: function get() {
    return _font_weights.FontWeightProps;
  }
});
Object.defineProperty(exports, "FontWeightRamp", {
  enumerable: true,
  get: function get() {
    return _font_weights.FontWeightRamp;
  }
});
Object.defineProperty(exports, "LineHeights", {
  enumerable: true,
  get: function get() {
    return _line_heights.LineHeights;
  }
});
Object.defineProperty(exports, "LineHeightProps", {
  enumerable: true,
  get: function get() {
    return _line_heights.LineHeightProps;
  }
});
Object.defineProperty(exports, "LineHeightRamp", {
  enumerable: true,
  get: function get() {
    return _line_heights.LineHeightRamp;
  }
});
Object.defineProperty(exports, "textDecoration", {
  enumerable: true,
  get: function get() {
    return _text_decoration.textDecoration;
  }
});
Object.defineProperty(exports, "textTransform", {
  enumerable: true,
  get: function get() {
    return _text_transform.textTransform;
  }
});

var _font_families = require("./font_families");

var _font_sizes = require("./font_sizes");

var _font_weights = require("./font_weights");

var _line_heights = require("./line_heights");

var _text_decoration = require("./text_decoration");

var _text_transform = require("./text_transform");
