// @looker/components/lib/Form/Inputs/Checkbox/Checkbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Checkbox = void 0;

var _noop = _interopRequireDefault(require("lodash/noop"));

var _pick = _interopRequireDefault(require("lodash/pick"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _designTokens = require("@looker/design-tokens");

var _isUndefined = _interopRequireDefault(require("lodash/isUndefined"));

var _InputProps = require("../InputProps");

var _InputText = require("../InputText");

var _CheckMark = require("./CheckMark");

var _CheckMarkMixed = require("./CheckMarkMixed");

var _FauxCheckbox = require("./FauxCheckbox");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var CheckboxLayout = (0, _react.forwardRef)((props, ref) => {
  var {
    className,
    checked,
    defaultChecked,
    onChange,
    readOnly,
    validationType
  } = props,
      restProps = _objectWithoutProperties(props, ["className", "checked", "defaultChecked", "onChange", "readOnly", "validationType"]);

  var [isChecked, setIsChecked] = (0, _react.useState)(!!defaultChecked);
  var handleClick = readOnly ? undefined : e => {
    if ((0, _isUndefined.default)(checked)) {
      setIsChecked(!isChecked);
    }

    if (onChange) {
      onChange(e);
    }
  };
  (0, _react.useEffect)(() => {
    if (!(0, _isUndefined.default)(checked)) {
      setIsChecked(checked);
    }
  }, [checked]);
  return _react.default.createElement("div", {
    className: className
  }, _react.default.createElement("input", _extends({
    type: "checkbox"
  }, (0, _pick.default)(restProps, _InputProps.inputPropKeys), {
    checked: !!isChecked,
    "aria-checked": checked,
    "aria-invalid": validationType === 'error' ? 'true' : undefined,
    onClick: handleClick,
    onChange: _noop.default,
    ref: ref
  })), _react.default.createElement(_FauxCheckbox.FauxCheckbox, null, checked === 'mixed' ? _react.default.createElement(_CheckMarkMixed.CheckMarkMixed, null) : _react.default.createElement(_CheckMark.CheckMark, null)));
});
CheckboxLayout.displayName = 'CheckboxLayout';
var Checkbox = (0, _styledComponents.default)(CheckboxLayout).withConfig({
  displayName: "Checkbox",
  componentId: "sc-9j2vap-0"
})(["", " ", " height:1rem;position:relative;width:1rem;input{cursor:", ";height:100%;opacity:0;position:absolute;width:100%;z-index:1;}input + ", "{", "}input:checked + ", "{background-color:", ";border-color:", ";}input:not(:checked) + ", "{color:", ";}input:focus + ", "{border-color:", ";box-shadow:0 0 0 2px ", ";outline:none;}input:disabled + ", "{background:", ";border-color:", ";color:", ";}input:disabled:not(:checked) + ", "{color:transparent;}"], _designTokens.reset, _designTokens.space, (_ref) => {
  var {
    readOnly,
    disabled
  } = _ref;
  return readOnly || disabled ? 'not-allowed' : undefined;
}, _FauxCheckbox.FauxCheckbox, _InputText.inputTextValidation, _FauxCheckbox.FauxCheckbox, (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.colors.key;
}, (_ref3) => {
  var {
    theme
  } = _ref3;
  return theme.colors.key;
}, _FauxCheckbox.FauxCheckbox, (_ref4) => {
  var {
    theme
  } = _ref4;
  return theme.colors.keyText;
}, _FauxCheckbox.FauxCheckbox, (_ref5) => {
  var {
    theme
  } = _ref5;
  return theme.colors.keyFocus;
}, (_ref6) => {
  var {
    theme
  } = _ref6;
  return theme.colors.keyAccent;
}, _FauxCheckbox.FauxCheckbox, (_ref7) => {
  var {
    theme
  } = _ref7;
  return theme.colors.ui1;
}, (_ref8) => {
  var {
    theme
  } = _ref8;
  return theme.colors.ui2;
}, (_ref9) => {
  var {
    theme
  } = _ref9;
  return theme.colors.text5;
}, _FauxCheckbox.FauxCheckbox);
exports.Checkbox = Checkbox;
