// @looker/design-tokens/lib/utils/color/stateful.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateStatefulColors = exports.generatePressed = exports.generateInteractive = void 0;

var _polished = require("polished");

var _blend = require("./blend");

var generateInteractive = color => (0, _polished.lighten)(0.04, color);

exports.generateInteractive = generateInteractive;

var generatePressed = color => (0, _polished.darken)(0.07, color);

exports.generatePressed = generatePressed;

var generateStatefulColor = (background, color) => {
  return {
    subtle: (0, _blend.mixScaledColors)(10, color, background),
    accent: (0, _blend.mixScaledColors)(16, color, background),
    focus: (0, _blend.mixScaledColors)(60, color, background),
    interactive: generateInteractive(color),
    pressed: generatePressed(color),
    text: background,
    border: color
  };
};

var generateStatefulColors = colors => {
  var {
    background,
    key,
    critical,
    neutral
  } = colors;
  var keyColors = generateStatefulColor(background, key);
  var criticalColors = generateStatefulColor(background, critical);
  var neutralColors = generateStatefulColor(background, neutral);
  return {
    keySubtle: keyColors.subtle,
    keyAccent: keyColors.accent,
    keyFocus: keyColors.focus,
    keyInteractive: keyColors.interactive,
    keyPressed: keyColors.pressed,
    keyText: keyColors.text,
    keyBorder: keyColors.border,
    criticalSubtle: criticalColors.subtle,
    criticalAccent: criticalColors.accent,
    criticalFocus: criticalColors.focus,
    criticalInteractive: criticalColors.interactive,
    criticalPressed: criticalColors.pressed,
    criticalText: criticalColors.text,
    criticalBorder: criticalColors.border,
    neutralSubtle: neutralColors.subtle,
    neutralAccent: neutralColors.accent,
    neutralFocus: neutralColors.focus,
    neutralInteractive: neutralColors.interactive,
    neutralPressed: neutralColors.pressed,
    neutralText: neutralColors.text,
    neutralBorder: neutralColors.border
  };
};

exports.generateStatefulColors = generateStatefulColors;
