// @looker/design-tokens/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  palette: true,
  prismTheme: true,
  generateThemeFromCoreColors: true,
  intentUIBlend: true,
  uiTransparencyBlend: true,
  generateIntentShade: true,
  generatePressed: true,
  generateInteractive: true
};
Object.defineProperty(exports, "prismTheme", {
  enumerable: true,
  get: function get() {
    return _prismTheme.prismTheme;
  }
});
Object.defineProperty(exports, "generateThemeFromCoreColors", {
  enumerable: true,
  get: function get() {
    return _theme2.generateThemeFromCoreColors;
  }
});
Object.defineProperty(exports, "intentUIBlend", {
  enumerable: true,
  get: function get() {
    return _blend.intentUIBlend;
  }
});
Object.defineProperty(exports, "uiTransparencyBlend", {
  enumerable: true,
  get: function get() {
    return _blend.uiTransparencyBlend;
  }
});
Object.defineProperty(exports, "generateIntentShade", {
  enumerable: true,
  get: function get() {
    return _blend.generateIntentShade;
  }
});
Object.defineProperty(exports, "generatePressed", {
  enumerable: true,
  get: function get() {
    return _stateful.generatePressed;
  }
});
Object.defineProperty(exports, "generateInteractive", {
  enumerable: true,
  get: function get() {
    return _stateful.generateInteractive;
  }
});
exports.palette = void 0;

var palette = _interopRequireWildcard(require("./tokens/color/palette"));

exports.palette = palette;

var _system = require("./system");

Object.keys(_system).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _system[key];
    }
  });
});

var _theme = require("./theme");

Object.keys(_theme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _theme[key];
    }
  });
});

var _GlobalStyle = require("./GlobalStyle");

Object.keys(_GlobalStyle).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _GlobalStyle[key];
    }
  });
});

var _prismTheme = require("./prismTheme");

var _animations = require("./utils/animations");

Object.keys(_animations).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _animations[key];
    }
  });
});

var _omit = require("./utils/omit");

Object.keys(_omit).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _omit[key];
    }
  });
});

var _theme2 = require("./utils/theme");

var _blend = require("./utils/color/blend");

var _stateful = require("./utils/color/stateful");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
