// @looker/components/lib/Dialog/Backdrop.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Backdrop = void 0;

var _designTokens = require("@looker/design-tokens");

var _styledSystem = require("styled-system");

var _styledComponents = _interopRequireDefault(require("styled-components"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Backdrop = _styledComponents.default.div.attrs(props => ({
  backgroundColor: props.visible ? props.backgroundColor : 'transparent'
})).withConfig({
  displayName: "Backdrop",
  componentId: "sc-195exav-0"
})(["", " ", " ", " bottom:0;cursor:default;left:0;opacity:", ";position:fixed;right:0;top:0;transition:opacity ", ";&.entering,&.exiting{opacity:0.01;}"], _designTokens.reset, _styledSystem.color, props => props.inlineStyle, props => props.opacity || 0.4, props => props.theme.transitions.durationSimple);

exports.Backdrop = Backdrop;
Backdrop.defaultProps = {
  backgroundColor: 'ui2',
  opacity: 0.6,
  visible: true
};
