// @looker/components/lib/Tree/TreeGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeGroup = exports.TreeGroupLabel = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _designTokens = require("@looker/design-tokens");

var _Accordion = require("../Accordion");

var _TreeItem = require("./TreeItem");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TreeGroupLayout = (_ref) => {
  var {
    children,
    className,
    label
  } = _ref;
  return _react.default.createElement("div", {
    className: className
  }, _react.default.createElement(TreeGroupLabel, null, label), children);
};

var TreeGroupLabel = _styledComponents.default.div.withConfig({
  displayName: "TreeGroup__TreeGroupLabel",
  componentId: "sc-1ylforq-0"
})(["border:1px transparent solid;font-size:", ";font-weight:", ";padding:", ";"], (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.fontSizes.xxsmall;
}, (_ref3) => {
  var {
    theme
  } = _ref3;
  return theme.fontWeights.semiBold;
}, (_ref4) => {
  var {
    theme: {
      space
    }
  } = _ref4;
  return "".concat(space.xsmall, " ").concat(space.xxsmall);
});

exports.TreeGroupLabel = TreeGroupLabel;
var TreeGroup = (0, _styledComponents.default)(TreeGroupLayout).withConfig({
  displayName: "TreeGroup",
  componentId: "sc-1ylforq-1"
})(["", ",", ",", "{", "}"], _TreeItem.TreeItemLabel, TreeGroupLabel, _Accordion.AccordionDisclosure, _designTokens.color);
exports.TreeGroup = TreeGroup;
