// @looker/components/lib/Avatar/AvatarUser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvatarUser = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Avatar = require("./Avatar");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AvatarLayout = (_ref) => {
  var {
    color,
    className,
    user,
    size
  } = _ref;
  var firstInitial = user && user.first_name && user.first_name[0];
  var lastInitial = user && user.last_name && user.last_name[0];
  return _react.default.createElement("div", {
    className: className
  }, _react.default.createElement(AvatarInitials, {
    color: color
  }, size === 'xxsmall' ? "".concat(firstInitial) : "".concat(firstInitial).concat(lastInitial)), user && user.avatar_url && _react.default.createElement(AvatarPhoto, {
    color: color,
    type: "image/png",
    data: user.avatar_url
  }));
};

var AvatarPhoto = _styledComponents.default.object.withConfig({
  displayName: "AvatarUser__AvatarPhoto",
  componentId: "ds3q2d-0"
})(["bottom:0;height:100%;left:0;position:absolute;right:0;top:0;width:100%;"]);

var AvatarInitials = _styledComponents.default.div.attrs(props => ({
  bg: props.color
})).withConfig({
  displayName: "AvatarUser__AvatarInitials",
  componentId: "ds3q2d-1"
})(["color:", ";"], (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.colors.keyText;
});

var AvatarUser = (0, _styledComponents.default)(AvatarLayout).attrs(props => ({
  size: props.size || 'small'
})).withConfig({
  displayName: "AvatarUser",
  componentId: "ds3q2d-2"
})(["", " background:currentColor;position:relative;"], _Avatar.avatarCSS);
exports.AvatarUser = AvatarUser;
AvatarUser.defaultProps = {
  color: 'key',
  size: 'small'
};
