// @looker/components/lib/Avatar/AvatarCombo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvatarCombo = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Icon = require("../Icon");

var _AvatarUser = require("./AvatarUser");

var _AvatarIcon = require("./AvatarIcon");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AvatarIconSecondary = (0, _styledComponents.default)(_AvatarIcon.AvatarIcon).withConfig({
  displayName: "AvatarCombo__AvatarIconSecondary",
  componentId: "gekgjm-0"
})([""]);

var AvatarLayout = (_ref) => {
  var {
    secondaryIcon,
    secondaryColor,
    secondaryBg,
    color,
    icon = 'User',
    user,
    className
  } = _ref;
  return _react.default.createElement("div", {
    className: className
  }, user ? _react.default.createElement(_AvatarUser.AvatarUser, {
    user: user,
    color: color
  }) : _react.default.createElement(_AvatarIcon.AvatarIcon, {
    color: color,
    icon: icon
  }), _react.default.createElement(AvatarIconSecondary, {
    bg: secondaryBg,
    color: secondaryColor,
    icon: secondaryIcon
  }));
};

var AvatarCombo = (0, _styledComponents.default)(AvatarLayout).withConfig({
  displayName: "AvatarCombo",
  componentId: "gekgjm-1"
})(["height:40px;position:relative;width:40px;", "{bottom:-4px;height:20px;position:absolute;right:-4px;width:20px;", "{height:14px;width:14px;}}"], AvatarIconSecondary, _Icon.Icon);
exports.AvatarCombo = AvatarCombo;
