// @looker/components/lib/Form/Inputs/AdvancedInputControls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdvancedInputControls = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _flatMap = _interopRequireDefault(require("lodash/flatMap"));

var _tail = _interopRequireDefault(require("lodash/tail"));

var _compact = _interopRequireDefault(require("lodash/compact"));

var _Icon = require("../../Icon");

var _InputSearchControls = require("./InputSearch/InputSearchControls");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var intersperseDivider = children => (0, _tail.default)((0, _flatMap.default)(children, (child, i) => [_react.default.createElement(SearchControlDivider, {
  key: i
}), child]));

var AdvancedInputControls = (_ref) => {
  var {
    validationType,
    renderSearchControls,
    onClear,
    disabled,
    isVisibleOptions,
    summary,
    hasOptions = true
  } = _ref;
  var children = intersperseDivider((0, _compact.default)([validationType === 'error' && _react.default.createElement(_Icon.Icon, {
    key: "warning",
    name: "CircleInfo",
    size: 20,
    color: "critical",
    mr: "xxsmall"
  }), renderSearchControls && _react.default.createElement(_InputSearchControls.InputSearchControls, {
    key: "search-controls",
    onClear: onClear,
    showClear: true,
    disabled: disabled,
    summary: summary
  }), hasOptions && _react.default.createElement(_Icon.Icon, {
    key: "list-caret",
    name: isVisibleOptions ? 'CaretUp' : 'CaretDown',
    size: 18,
    color: disabled ? 'text6' : 'text4',
    mr: "xsmall"
  })]));
  return _react.default.createElement(SearchControlGrid, null, children);
};

exports.AdvancedInputControls = AdvancedInputControls;

var SearchControlGrid = _styledComponents.default.div.withConfig({
  displayName: "AdvancedInputControls__SearchControlGrid",
  componentId: "sc-1e7uo3l-0"
})(["align-items:center;display:grid;grid-auto-flow:column dense;grid-gap:", ";justify-items:center;max-height:1.9rem;"], (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.space.xxsmall;
});

var SearchControlDivider = _styledComponents.default.div.withConfig({
  displayName: "AdvancedInputControls__SearchControlDivider",
  componentId: "sc-1e7uo3l-1"
})(["background:", ";height:70%;width:1px;"], (_ref3) => {
  var {
    theme
  } = _ref3;
  return theme.colors.ui2;
});
