// @looker/components/lib/Tooltip/TooltipContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipContent = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Link = require("../Link");

var _Text = require("../Text");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TooltipContent = (0, _styledComponents.default)(_Text.Paragraph).attrs(props => ({
  fontSize: 'xsmall',
  lineHeight: 'xsmall',
  m: 'none',
  maxWidth: props.width,
  p: 'xsmall',
  width: 'auto'
})).withConfig({
  displayName: "TooltipContent",
  componentId: "sc-1fmi5qh-0"
})(["hyphens:auto;overflow-wrap:anywhere;text-transform:none;white-space:normal;word-break:break-word;", "{color:", ";text-decoration:underline;&:focus,&:hover{color:", ";}&:active{color:", ";}}"], _Link.Link, props => props.theme.colors.keyAccent, props => props.theme.colors.keySubtle, props => props.theme.colors.keyText);
exports.TooltipContent = TooltipContent;
TooltipContent.defaultProps = {
  textAlign: 'center',
  width: '16rem'
};
