// @looker/components/lib/Tooltip/tooltip.test.helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertClosed = exports.assertOpen = exports.assertOverlayState = exports.mouseEventSimulator = exports.SimpleContentSFC = exports.SimpleContent = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SimpleContent = _react.default.createElement("div", null, "simple content");

exports.SimpleContent = SimpleContent;

var SimpleContentSFC = () => SimpleContent;

exports.SimpleContentSFC = SimpleContentSFC;
var mouseEventSimulator = {
  currentTarget: {},
  relatedTarget: document.body
};
exports.mouseEventSimulator = mouseEventSimulator;

var assertOverlayState = function assertOverlayState(overlay, content) {
  var open = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  expect(overlay.contains(content)).toEqual(open);
};

exports.assertOverlayState = assertOverlayState;

var assertOpen = overlay => assertOverlayState(overlay, SimpleContent);

exports.assertOpen = assertOpen;

var assertClosed = overlay => assertOverlayState(overlay, SimpleContent, false);

exports.assertClosed = assertClosed;
