// @looker/components/lib/Form/Inputs/Select/SelectMulti.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectMulti = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Combobox = require("../Combobox");

var _SelectOptions = require("./SelectOptions");

var _options = require("./utils/options");

var _useWindowedOptions = require("./utils/useWindowedOptions");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var SelectMultiComponent = (0, _react.forwardRef)((_ref, ref) => {
  var {
    options,
    disabled,
    isFilterable = false,
    placeholder,
    onFilter,
    onChange,
    values,
    defaultValues,
    noOptionsLabel,
    'aria-describedby': ariaDescribedby,
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledby,
    indicator,
    listLayout,
    validationType,
    windowedOptions: windowedOptionsProp,
    closeOnSelect = false,
    showCreate = false,
    formatCreateLabel,
    removeOnBackspace = true,
    freeInput = false
  } = _ref,
      props = _objectWithoutProperties(_ref, ["options", "disabled", "isFilterable", "placeholder", "onFilter", "onChange", "values", "defaultValues", "noOptionsLabel", "aria-describedby", "aria-label", "aria-labelledby", "indicator", "listLayout", "validationType", "windowedOptions", "closeOnSelect", "showCreate", "formatCreateLabel", "removeOnBackspace", "freeInput"]);

  var optionValues = (0, _options.getOptions)(values, options);
  var defaultOptionValues = (0, _options.getOptions)(defaultValues, options);

  function handleChange(options) {
    var newValues = options && options.map(option => option.value);
    onChange && onChange(newValues);
    onFilter && onFilter('');
  }

  function handleInputChange(value) {
    onFilter && onFilter(value);
  }

  function handleClose() {
    onFilter && onFilter('');
  }

  var ariaProps = {
    'aria-describedby': ariaDescribedby,
    'aria-invalid': validationType === 'error',
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledby
  };
  var windowedOptions = (0, _useWindowedOptions.useShouldWindowOptions)(options, windowedOptionsProp);
  return _react.default.createElement(_Combobox.ComboboxMulti, _extends({}, props, {
    values: optionValues,
    defaultValues: defaultOptionValues,
    onChange: handleChange,
    onClose: handleClose
  }), _react.default.createElement(_Combobox.ComboboxMultiInput, _extends({}, ariaProps, {
    disabled: disabled,
    placeholder: placeholder,
    removeOnBackspace: removeOnBackspace,
    validationType: validationType,
    autoComplete: false,
    readOnly: !isFilterable,
    onInputChange: handleInputChange,
    selectOnClick: isFilterable,
    freeInput: freeInput,
    ref: ref
  })), !disabled && _react.default.createElement(_Combobox.ComboboxMultiList, _extends({
    persistSelection: true,
    closeOnSelect: closeOnSelect,
    windowedOptions: windowedOptions,
    indicator: indicator
  }, ariaProps, listLayout), _react.default.createElement(_SelectOptions.SelectOptions, {
    isMulti: true,
    options: options,
    windowedOptions: windowedOptions,
    isFilterable: isFilterable,
    noOptionsLabel: noOptionsLabel,
    showCreate: showCreate,
    formatCreateLabel: formatCreateLabel
  })));
});
SelectMultiComponent.displayName = 'SelectMultiComponent';
var SelectMulti = (0, _styledComponents.default)(SelectMultiComponent).withConfig({
  displayName: "SelectMulti",
  componentId: "sc-176rl8-0"
})([""]);
exports.SelectMulti = SelectMulti;
