// @looker/components/lib/Accordion/AccordionDisclosureGrid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AccordionDisclosureGrid = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Icon = require("../Icon");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Indicator = _styledComponents.default.div.withConfig({
  displayName: "AccordionDisclosureGrid__Indicator",
  componentId: "ux3ubb-0"
})(["align-items:center;display:flex;grid-area:indicator;justify-content:center;"]);

var Label = _styledComponents.default.div.withConfig({
  displayName: "AccordionDisclosureGrid__Label",
  componentId: "ux3ubb-1"
})(["grid-area:children;"]);

var AccordionDisclosureGridLayout = (_ref) => {
  var {
    children,
    className,
    isOpen,
    indicatorIcons,
    indicatorSize
  } = _ref;
  return _react.default.createElement("div", {
    className: className
  }, _react.default.createElement(Indicator, null, _react.default.createElement(_Icon.Icon, {
    name: isOpen ? indicatorIcons.open : indicatorIcons.close,
    size: indicatorSize
  })), _react.default.createElement(Label, null, children));
};

var AccordionDisclosureGrid = (0, _styledComponents.default)(AccordionDisclosureGridLayout).withConfig({
  displayName: "AccordionDisclosureGrid",
  componentId: "ux3ubb-2"
})(["align-items:center;display:grid;grid-gap:", ";grid-template-areas:", ";grid-template-columns:", ";width:100%;"], (_ref2) => {
  var {
    indicatorGap,
    theme
  } = _ref2;
  return theme.space[indicatorGap];
}, (_ref3) => {
  var {
    indicatorPosition
  } = _ref3;
  return indicatorPosition === 'left' ? '"indicator children"' : '"children indicator"';
}, (_ref4) => {
  var {
    indicatorPosition,
    indicatorSize,
    theme
  } = _ref4;
  return indicatorPosition === 'left' ? theme.space[indicatorSize] + ' 1fr' : '1fr ' + theme.space[indicatorSize];
});
exports.AccordionDisclosureGrid = AccordionDisclosureGrid;
