// @looker/components/lib/ActionList/ActionListItemColumn.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionListItemColumn = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ = require("../");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ActionListItemColumnInnerLayout = _styledComponents.default.div.withConfig({
  displayName: "ActionListItemColumn__ActionListItemColumnInnerLayout",
  componentId: "d6ni17-0"
})([""]);

var ActionListItemColumnLayout = (_ref) => {
  var {
    children,
    detail,
    className,
    indicator
  } = _ref;
  return _react.default.createElement("div", {
    className: className
  }, indicator && _react.default.createElement(_.Flex, {
    justifyContent: "center",
    alignItems: "center",
    mr: "xsmall"
  }, indicator), _react.default.createElement(ActionListItemColumnInnerLayout, null, children, detail && _react.default.createElement(_.Paragraph, {
    fontSize: "xsmall",
    variant: "subdued",
    truncate: true
  }, detail)));
};

var ActionListItemColumn = (0, _styledComponents.default)(ActionListItemColumnLayout).withConfig({
  displayName: "ActionListItemColumn",
  componentId: "d6ni17-1"
})(["color:", ";display:", ";font-size:", ";overflow:hidden;word-break:break-all;", "{display:flex;flex-direction:column;justify-content:center;}"], (_ref2) => {
  var {
    theme
  } = _ref2;
  return theme.colors.text2;
}, (_ref3) => {
  var {
    indicator
  } = _ref3;
  return indicator && 'flex';
}, (_ref4) => {
  var {
    theme
  } = _ref4;
  return theme.fontSizes.xsmall;
}, ActionListItemColumnInnerLayout);
exports.ActionListItemColumn = ActionListItemColumn;
