// @looker/components/lib/ActionList/ActionListRow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionListRow = exports.ActionListRowColumns = void 0;

var _pick = _interopRequireDefault(require("lodash/pick"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ActionListCheckbox = require("./ActionListCheckbox");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ActionListRowColumns = _styledComponents.default.div.withConfig({
  displayName: "ActionListRow__ActionListRowColumns",
  componentId: "yaf080-0"
})([""]);

exports.ActionListRowColumns = ActionListRowColumns;

var ActionListRowSupplementary = _styledComponents.default.div.withConfig({
  displayName: "ActionListRow__ActionListRowSupplementary",
  componentId: "yaf080-1"
})([""]);

var ActionListRowLayout = (0, _react.forwardRef)((props, ref) => {
  var {
    className,
    hasCheckbox,
    children,
    onClick,
    onKeyDown,
    secondary,
    tabIndex
  } = props;
  return _react.default.createElement("div", {
    ref: ref,
    className: className,
    onKeyDown: onKeyDown,
    tabIndex: tabIndex
  }, hasCheckbox && _react.default.createElement(_ActionListCheckbox.ActionListCheckbox, (0, _pick.default)(props, _ActionListCheckbox.checkListProps)), _react.default.createElement(ActionListRowColumns, {
    onClick: onClick
  }, children), _react.default.createElement(ActionListRowSupplementary, null, secondary));
});
ActionListRowLayout.displayName = 'ActionListRowLayout';
var ActionListRow = (0, _styledComponents.default)(ActionListRowLayout).withConfig({
  displayName: "ActionListRow",
  componentId: "yaf080-2"
})(["background:", ";display:flex;&:focus,&:hover{box-shadow:", ";cursor:", ";outline:none;position:relative;}", "{flex-grow:1;}", "{align-items:center;display:flex;flex-basis:2.5rem;flex-shrink:0;}"], (_ref) => {
  var {
    disabled,
    theme
  } = _ref;
  return disabled && theme.colors.ui1;
}, (_ref2) => {
  var {
    theme,
    supportsRaised,
    onClick
  } = _ref2;
  return supportsRaised && onClick && theme.shadows[2];
}, (_ref3) => {
  var {
    onClick
  } = _ref3;
  return onClick && 'pointer';
}, ActionListRowColumns, ActionListRowSupplementary);
exports.ActionListRow = ActionListRow;
ActionListRow.defaultProps = {
  supportsRaised: true
};
