// @looker/components/lib/ActionList/ActionListCheckbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionListCheckbox = exports.actionListCheckboxWidth = exports.checkListProps = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Form = require("../Form");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var checkListProps = ['checked', 'disabled', 'onChange'];
exports.checkListProps = checkListProps;
var actionListCheckboxWidth = '2.75rem';
exports.actionListCheckboxWidth = actionListCheckboxWidth;

var ActionListCheckboxLayout = (_ref) => {
  var {
    onChange,
    checked,
    disabled,
    className
  } = _ref;

  var handleOnChange = event => {
    var isNotFromChild = event.currentTarget === event.target;
    isNotFromChild && !disabled && onChange && onChange();
  };

  var handleOnKeyDown = event => {
    if (event.keyCode === 13) {
      event.currentTarget.click();
    }
  };

  return _react.default.createElement("div", {
    onClick: handleOnChange,
    className: className
  }, _react.default.createElement(_Form.Checkbox, {
    disabled: disabled,
    checked: checked,
    onChange: handleOnChange,
    onKeyDown: handleOnKeyDown
  }));
};

var ActionListCheckbox = (0, _styledComponents.default)(ActionListCheckboxLayout).withConfig({
  displayName: "ActionListCheckbox",
  componentId: "sc-17hye4r-0"
})(["align-items:center;cursor:", ";display:flex;flex-basis:", ";flex-shrink:0;justify-content:center;"], (_ref2) => {
  var {
    disabled
  } = _ref2;
  return disabled ? 'not-allowed' : 'default';
}, actionListCheckboxWidth);
exports.ActionListCheckbox = ActionListCheckbox;
