// @looker/components/lib/Form/Inputs/Combobox/ComboboxOption.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComboboxOptionTextInternal = ComboboxOptionTextInternal;
exports.ComboboxOptionText = exports.comboboxOptionDefaultProps = exports.ComboboxOption = exports.comboboxOptionStyle = exports.ComboboxOptionWrapper = void 0;

var _designTokens = require("@looker/design-tokens");

var _omit = _interopRequireDefault(require("lodash/omit"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _Icon = require("../../../Icon");

var _Text = require("../../../Text");

var _utils = require("../../../utils");

var _makeHash = require("./utils/makeHash");

var _ComboboxContext = require("./ComboboxContext");

var _getComboboxText = require("./utils/getComboboxText");

var _useOptionEvents = require("./utils/useOptionEvents");

var _useOptionStatus = require("./utils/useOptionStatus");

var _useAddOptionToContext = require("./utils/useAddOptionToContext");

var _ComboboxOptionIndicator = require("./ComboboxOptionIndicator");

var _useOptionScroll = require("./utils/useOptionScroll");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ComboboxOptionWrapper = (0, _react.forwardRef)((_ref, forwardedRef) => {
  var {
    children,
    label,
    value
  } = _ref,
      rest = _objectWithoutProperties(_ref, ["children", "label", "value"]);

  return _react.default.createElement(_ComboboxContext.OptionContext.Provider, {
    value: {
      label,
      value
    }
  }, _react.default.createElement("li", _extends({}, (0, _omit.default)(rest, 'color'), {
    ref: forwardedRef,
    id: String((0, _makeHash.makeHash)(value)),
    role: "option",
    tabIndex: -1
  }), children));
});
exports.ComboboxOptionWrapper = ComboboxOptionWrapper;
ComboboxOptionWrapper.displayName = 'ComboboxOptionWrapper';
var ComboboxOptionInternal = (0, _react.forwardRef)((_ref2, forwardedRef) => {
  var {
    children,
    indicator,
    highlightText = true,
    scrollIntoView
  } = _ref2,
      props = _objectWithoutProperties(_ref2, ["children", "indicator", "highlightText", "scrollIntoView"]);

  var {
    label,
    value
  } = props;
  (0, _useAddOptionToContext.useAddOptionToContext)(_ComboboxContext.ComboboxContext, value, label, scrollIntoView);
  var optionEvents = (0, _useOptionEvents.useOptionEvents)(props, _ComboboxContext.ComboboxContext);
  var {
    isActive,
    isSelected
  } = (0, _useOptionStatus.useOptionStatus)(_ComboboxContext.ComboboxContext, value);
  var scrollRef = (0, _useOptionScroll.useOptionScroll)(_ComboboxContext.ComboboxContext, value, label, scrollIntoView, isActive);
  var ref = (0, _utils.useForkedRef)(scrollRef, forwardedRef);
  return _react.default.createElement(ComboboxOptionWrapper, _extends({}, props, optionEvents, {
    ref: ref,
    "aria-selected": isActive
  }), _react.default.createElement(_ComboboxOptionIndicator.ComboboxOptionIndicator, {
    indicator: indicator,
    isActive: isActive,
    isSelected: isSelected
  }, isSelected && _react.default.createElement(_Icon.Icon, {
    name: "Check",
    size: "xsmall",
    mr: 0
  })), children || _react.default.createElement(ComboboxOptionText, {
    highlightText: highlightText
  }));
});
ComboboxOptionInternal.displayName = 'ComboboxOptionInternal';
var comboboxOptionStyle = (0, _styledComponents.css)(["", " ", " ", " ", " ", " ", " align-items:stretch;cursor:default;outline:none;&[aria-selected='true']{background-color:", ";color:", ";}"], _designTokens.reset, _designTokens.color, _designTokens.flexbox, _designTokens.layout, _designTokens.space, _designTokens.typography, props => props.theme.colors.keySubtle, props => props.theme.colors.keyPressed);
exports.comboboxOptionStyle = comboboxOptionStyle;
var ComboboxOption = (0, _styledComponents.default)(ComboboxOptionInternal).withConfig({
  displayName: "ComboboxOption",
  componentId: "w994y4-0"
})(["", ""], comboboxOptionStyle);
exports.ComboboxOption = ComboboxOption;
var comboboxOptionDefaultProps = {
  color: 'text2',
  display: 'flex',
  fontSize: 'small',
  px: 'xsmall',
  py: 'xxsmall'
};
exports.comboboxOptionDefaultProps = comboboxOptionDefaultProps;
ComboboxOption.defaultProps = comboboxOptionDefaultProps;

function ComboboxOptionTextInternal(_ref3) {
  var {
    highlightText = true
  } = _ref3,
      props = _objectWithoutProperties(_ref3, ["highlightText"]);

  var context = (0, _react.useContext)(_ComboboxContext.ComboboxContext);
  var contextMulti = (0, _react.useContext)(_ComboboxContext.ComboboxMultiContext);
  var contextToUse = context.transition ? context : contextMulti;
  var {
    data
  } = contextToUse;
  var {
    inputValue
  } = data;
  var contextOption = data.option;
  var option = (0, _react.useContext)(_ComboboxContext.OptionContext);
  var text = (0, _getComboboxText.getComboboxText)(option);

  if (!highlightText || !inputValue || inputValue === '' || inputValue === (0, _getComboboxText.getComboboxText)(contextOption)) {
    return _react.default.createElement("span", props, text);
  }

  return _react.default.createElement("span", props, _react.default.createElement(_Text.ReplaceText, {
    match: inputValue,
    replace: (str, index) => _react.default.createElement(_Text.Text, {
      fontWeight: "semiBold",
      fontSize: "small",
      textDecoration: "underline",
      key: index
    }, str)
  }, text));
}

var ComboboxOptionText = (0, _styledComponents.default)(ComboboxOptionTextInternal).withConfig({
  displayName: "ComboboxOption__ComboboxOptionText",
  componentId: "w994y4-1"
})([""]);
exports.ComboboxOptionText = ComboboxOptionText;
