// @looker/components/lib/Form/Fields/FieldTextArea/FieldTextArea.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FieldTextArea = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _utils = require("../../../utils");

var _Form = require("../../Form");

var _TextArea = require("../../Inputs/TextArea");

var _Field = require("../Field");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var FieldTextAreaComponent = (_ref) => {
  var props = _extends({}, _ref);

  var id = (0, _utils.useID)(props.id);
  var validationMessage = (0, _Form.useFormContext)(props);
  return _react.default.createElement(_Field.Field, _extends({}, (0, _Field.pickFieldProps)(props), {
    id: id,
    validationMessage: validationMessage
  }), _react.default.createElement(_TextArea.TextArea, _extends({}, (0, _Field.omitFieldProps)(props), {
    id: id,
    "aria-describedby": "".concat(id, "-describedby"),
    validationType: validationMessage && validationMessage.type
  })));
};

FieldTextAreaComponent.displayName = 'FieldTextAreaComponent';
var FieldTextArea = (0, _styledComponents.default)(FieldTextAreaComponent).withConfig({
  displayName: "FieldTextArea",
  componentId: "sc-1nmc8fy-0"
})([""]);
exports.FieldTextArea = FieldTextArea;
