// @looker/components/lib/Layout/Space/Space.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Space = exports.spaceCSS = exports.defaultGap = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _styledSystem = require("styled-system");

var _designTokens = require("@looker/design-tokens");

var _simple = require("../utils/simple");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var defaultGap = 'medium';
exports.defaultGap = defaultGap;
var spaceCSS = (0, _styledComponents.css)(["", " ", " display:flex;"], _simple.simpleLayoutCSS, _designTokens.flexbox);
exports.spaceCSS = spaceCSS;

var flexGap = (_ref) => {
  var {
    gap = defaultGap,
    reverse
  } = _ref;
  return (0, _styledComponents.css)(["@supports (-moz-appearance:none){gap:", ";}@supports not (-moz-appearance:none){&& > *{margin-right:", ";}", "}"], (_ref2) => {
    var {
      theme: {
        space
      }
    } = _ref2;
    return space[gap];
  }, (_ref3) => {
    var {
      theme: {
        space
      }
    } = _ref3;
    return space[gap];
  }, (_ref4) => {
    var {
      theme: {
        space
      }
    } = _ref4;
    return reverse ? "&& > *:first-child { margin-right: ".concat(space.none, "; }") : "&& > *:last-child { margin-right: ".concat(space.none, "; }");
  });
};

var verticalAlign = (0, _styledSystem.variant)({
  prop: 'align',
  variants: {
    center: {
      alignItems: 'center'
    },
    end: {
      alignItems: 'flex-end'
    },
    start: {
      alignItems: 'flex-start'
    }
  }
});

var Space = _styledComponents.default.div.withConfig({
  displayName: "Space",
  componentId: "zsz5hl-0"
})(["", " ", " flex-direction:", ";", " ", " ", " ", ""], spaceCSS, (_ref5) => {
  var {
    stretch
  } = _ref5;
  return !stretch && verticalAlign;
}, (_ref6) => {
  var {
    reverse
  } = _ref6;
  return reverse ? 'row-reverse' : 'row';
}, (_ref7) => {
  var {
    around
  } = _ref7;
  return around && 'justify-content: space-around;';
}, (_ref8) => {
  var {
    between
  } = _ref8;
  return between && 'justify-content: space-between;';
}, (_ref9) => {
  var {
    evenly
  } = _ref9;
  return evenly && 'justify-content: space-evenly;';
}, (_ref10) => {
  var {
    around,
    between,
    evenly
  } = _ref10;
  return !around && !between && !evenly && flexGap;
});

exports.Space = Space;
Space.defaultProps = {
  alignItems: 'center',
  width: '100%'
};
