// @looker/components/lib/Layout/Space/SpaceVertical.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpaceVertical = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _styledSystem = require("styled-system");

var _Space = require("./Space");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var align = (0, _styledSystem.variant)({
  prop: 'align',
  variants: {
    center: {
      alignItems: 'center'
    },
    end: {
      alignItems: 'flex-end'
    },
    start: {
      alignItems: 'flex-start'
    },
    stretch: {
      alignItems: 'stretch'
    }
  }
});

var flexGap = (_ref) => {
  var {
    gap = _Space.defaultGap,
    reverse
  } = _ref;
  return (0, _styledComponents.css)(["@supports (-moz-appearance:none){gap:", ";}@supports not (-moz-appearance:none){&& > *{margin-top:", ";}", "}"], (_ref2) => {
    var {
      theme: {
        space
      }
    } = _ref2;
    return space[gap];
  }, (_ref3) => {
    var {
      theme: {
        space
      }
    } = _ref3;
    return space[gap];
  }, (_ref4) => {
    var {
      theme: {
        space
      }
    } = _ref4;
    return reverse ? "&& > *:last-child { margin-top: ".concat(space.none, "; }") : "&& > *:first-child { margin-top: ".concat(space.none, "; }");
  });
};

var SpaceVertical = _styledComponents.default.div.withConfig({
  displayName: "SpaceVertical",
  componentId: "tnj0rp-0"
})(["", " ", " ", " flex-direction:", ";"], _Space.spaceCSS, align, flexGap, (_ref5) => {
  var {
    reverse
  } = _ref5;
  return reverse ? 'column-reverse' : 'column';
});

exports.SpaceVertical = SpaceVertical;
SpaceVertical.defaultProps = {
  align: 'stretch',
  width: '100%'
};
